### PR TITLE
Add basic auth & response code check for deploy-from-url

### PR DESCRIPTION
**Pull request by radito3**
#### Description: 
Add support for basic authentication in deploy from URL feature.
Add check for response code when deploying from URL.

### CODE REVIEW

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 1]
```diff
@@ -159,12 +164,20 @@ private FileMetadata parseFileEntry(FileEntry fileEntry) {
...
+                                             .decode(fileUrl));
+        UriUtil.validateUrl(decodedUrl);
+        HttpClient client = buildHttpClient(decodedUrl);
 
         HttpResponse<InputStream> response = client.send(buildFetchFileRequest(decodedUrl), BodyHandlers.ofInputStream());
 
+        if (response.statusCode() / 100 != 2) {
+            try (InputStream is = response.body()) {
+                String error = IOUtils.toString(is, StandardCharsets.UTF_8);
+                throw new SLException("Error from remote MTAR endpoint: " + error);
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:177]: Extract in constant

Comment (IvanBorislavovDimitrov) [reply multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:177]: Maybe we should add resilience at some point in time

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 2]
```diff
@@ -181,21 +194,38 @@ private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String f
...
                          .version(HttpClient.Version.HTTP_1_1)
                          .connectTimeout(Duration.ofMinutes(10))
                          .followRedirects(Redirect.NORMAL)
+                         .authenticator(new Authenticator() {
+                             @Override
+                             protected PasswordAuthentication getPasswordAuthentication() {
+                                 if (decodedUrl.contains("@")) {
+                                     var uri = URI.create(decodedUrl);
+                                     var userInfo = uri.getUserInfo();
+                                     var userPassParts = userInfo.split(":");
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:208]: Do not split it! The password can contain ":", the username cannot. Use substring.

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 3]
```diff
@@ -181,21 +194,38 @@ private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String f
...
-    protected HttpClient buildHttpClient() {
+    protected HttpClient buildHttpClient(String decodedUrl) {
         return HttpClient.newBuilder()
-                         //ssl and authentication configuration can be done here
                          .version(HttpClient.Version.HTTP_1_1)
                          .connectTimeout(Duration.ofMinutes(10))
                          .followRedirects(Redirect.NORMAL)
+                         .authenticator(new Authenticator() {
+                             @Override
+                             protected PasswordAuthentication getPasswordAuthentication() {
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:204]: Extract this anonymous class

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 4]
```diff
@@ -181,21 +194,38 @@ private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String f
...
+                                 if (decodedUrl.contains("@")) {
+                                     var uri = URI.create(decodedUrl);
+                                     var userInfo = uri.getUserInfo();
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:207]: https://abv.com/one@two  -> NPE
I think first the URI should be created, then check if userInfo is null

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 5]
```diff
@@ -181,21 +194,38 @@ private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String f
...
-                          .uri(URI.create(decodedUrl))
-                          .timeout(Duration.ofMinutes(5))
-                          .build();
+        var builder = HttpRequest.newBuilder()
+                                 .GET()
+                                 .timeout(Duration.ofMinutes(5));
+        var uri = URI.create(decodedUrl);
+        if (decodedUrl.contains("@")) {
+            var userInfo = uri.getUserInfo();
+            builder.uri(URI.create(decodedUrl.replace(userInfo + "@", "")));
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:224]: Same as above
Extract in method and just return the formatted URI

File: multiapps-controller-web/src/test/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImplTest.java [hunk 6]
```diff
@@ -2,17 +2,15 @@
...
 import java.net.http.HttpHeaders;
 import java.net.http.HttpResponse;
 import java.net.http.HttpResponse.BodyHandlers;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import java.util.Random;
-import java.util.UUID;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/test/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImplTest.java:13]: NO

**Review by IvanBorislavovDimitrov — CHANGES_REQUESTED** [review]

File: multiapps-controller-web/src/test/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImplTest.java (hunk 6 above)
Comment (IvanBorislavovDimitrov) [reply multiapps-controller-web/src/test/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImplTest.java:13]: Is this removed?

**Review by IvanBorislavovDimitrov — APPROVED** [review]

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 7]
```diff
@@ -157,13 +166,16 @@ private FileMetadata parseFileEntry(FileEntry fileEntry) {
...
-        throws FileStorageException, IOException, InterruptedException {
-        UriUtil.validateUrl(fileUrl);
-        String decodedUrl = URLDecoder.decode(fileUrl, StandardCharsets.UTF_8);
-        HttpClient client = buildHttpClient();
+    private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String fileUrl) throws Exception {
+        String decodedUrl = new String(Base64.getUrlDecoder()
+                                             .decode(fileUrl));
+        if (!decodedUrl.startsWith(UriUtil.HTTPS_PROTOCOL)) {
+            throw new SLException(Messages.MTAR_ENDPOINT_NOT_SECURE);
+        }
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:174]: Move this check in the UriValidator

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java [hunk 8]
```diff
@@ -181,21 +193,59 @@ private FileEntry uploadFileFromUrl(String spaceGuid, String namespace, String f
...
+            }
+            return response;
+        });
+    }
+
+    protected HttpClient buildHttpClient(String decodedUrl) {
         return HttpClient.newBuilder()
-                         //ssl and authentication configuration can be done here
-                         .version(HttpClient.Version.HTTP_1_1)
+                         .version(HttpClient.Version.HTTP_2)
```
Comment (IvanBorislavovDimitrov) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:213]: Do all urls work with http 2?

Comment (radito3) [reply multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/FilesApiServiceImpl.java:213]: No, but the client will fallback to HTTP/1.1 if 2 doesn't work.

**Pull request merged**